extensions/gitpod/src/remoteConnector.ts

```diff
@@ -379,22 +396,83 @@ export default class LocalApp extends Disposable {
 		}
 	}
 
-	public async handleUri(uri: vscode.Uri) {
-		if (uri.path === LocalApp.AUTH_COMPLETE_PATH) {
-			this.logger.info('auth completed');
-			return;
+	private async getWorkspaceSSHDestination(workspaceId: string, gitpodHost: string): Promise<string> {
+		const session = await vscode.authentication.getSession(
+			'gitpod',
+			['function:getWorkspace', 'function:getOwnerToken', 'function:getLoggedInUser', 'resource:default'],
+			{ createIfNone: true }
+		);
+
+		const serviceUrl = new URL(gitpodHost);
+
+		const workspaceInfo = await withServerApi(session.accessToken, serviceUrl.toString(), service => service.server.getWorkspace(workspaceId), this.logger);
+		if (!workspaceInfo.latestInstance) {
+			throw new Error('no_running_instance');
 		}
-		this.logger.info('open workspace window: ' + uri.toString());
-		const params: SSHConnectionParams = JSON.parse(uri.query);
-		let resolvedConfig: LocalAppConfig | undefined;
-		try {
-			await vscode.window.withProgress({
-				location: vscode.ProgressLocation.Notification,
-				cancellable: true,
-				title: `Connecting to Gitpod workspace: ${params.workspaceId}`
-			}, async (_, token) => {
+
+		const workspaceUrl = new URL(workspaceInfo.latestInstance.ideUrl);
+
+		const sshHostKeyEndPoint = `https://${workspaceUrl.host}/_ssh/host_keys`;
+		const sshHostKeyResponse = await fetch(sshHostKeyEndPoint);
+		if (!sshHostKeyResponse.ok) {
+			// Gitpod SSH gateway not configured
+			throw new Error('no_ssh_gateway');
+		}
+
+		const sshHostKeys: { type: string, host_key: string }[] = await sshHostKeyResponse.json();
+
+		const ownerToken = await withServerApi(session.accessToken, serviceUrl.toString(), service => service.server.getOwnerToken(workspaceId), this.logger);
+
+		const sshDestInfo = {
+			user: `${workspaceId}#${ownerToken}`,
+			// See https://github.com/gitpod-io/gitpod/pull/9786 for reasoning about `.ssh` suffix
+			hostName: workspaceUrl.host.replace(workspaceId, `${workspaceId}.ssh`)
+		};
+
+		// Test ssh connection first
+		await new Promise<void>((resolve, reject) => {
+			const conn = new sshClient();
+			conn.on('ready', () => {
+				conn.end();
+				resolve();
+			}).on('error', err => {
+				reject(new SSHError(err));
+			}).connect({
+				host: sshDestInfo.hostName,
+				username: sshDestInfo.user,
+				authHandler(methodsLeft, _partialSuccess, callback) {
+					if (!methodsLeft) {
+						callback({
+							type: 'password',
+							username: workspaceId,
+							password: ownerToken
+						});
+					} else {
+						callback(false);
+					}
+				},
+				hostVerifier(hashedKey) {
```

Comment:
Hm, i hoped that this library updates known hosts file if it is successful as JB does. Could find another library or update this file ourself if connection is successful?